timemachine/fe/utils.py

```diff
@@ -250,3 +257,62 @@ def delta_U(from_idx, to_idx):
         delta_Us.append((fwd_delta_U, rev_delta_U))
 
     return np.array(delta_Us)
+
+
+def wrap_for_scipy_optimize(U, x0):
+    """Given a jax transformable function U and a configuration x0,
+    produce a flattened value_and_grad function compatible with the scipy.optimize L-BFGS-B implementation"""
+    shape = x0.shape
+
+    def wrapped(x):
+        v, g = value_and_grad(U)(x.reshape(shape))
+        return float(v), np.array(g.flatten(), dtype=np.float64)
+
+    return wrapped
+
+
+def minimize_energy(U_fxn, x0, verbose=True):
+    """using L-BFGS"""
+
+    fun = wrap_for_scipy_optimize(U_fxn, x0)
+    res = minimize(fun, x0.flatten(), jac=True, method="L-BFGS-B")
+    if verbose:
+        print(res)
+    return res.x.reshape(x0.shape)
+
+
+def partially_resample_configuration(U, x0, selection_mask, temperature=DEFAULT_TEMP, minimize_first=False):
+    r"""Approximately resample
+    x0[selection_mask] ~ p(x[selection_mask] | x0[!selection_mask])
+    where p(x) \propto exp(-U(x)/(BOLTZ * temperature))
+    using a few steps of Langevin dynamics
+    """
+    x0 = jnp.array(x0)
+    x0_subset = x0[selection_mask]
+
+    # prepare energy function defined on selected atoms only
+    def U_subset(x_subset):
+        return U(x0.at[selection_mask].set(x_subset))
+
+    # optionally minimize
+    if minimize_first:
+        x0_subset = minimize_energy(jit(U_subset), x0_subset)
```

Comment:
May be superseded by https://github.com/proteneer/timemachine/pull/821/files#diff-bb2d19446b7ea37b639b8b5a06a398d7be930ad81e47d67e143e2da183f07e1cR325-R343